src/Attr.cc

```diff
@@ -195,6 +195,33 @@ Attributes::Attributes(std::vector<AttrPtr> a,
 
 void Attributes::AddAttr(AttrPtr attr)
 	{
+	auto acceptable_duplicate_tag = [&](AttrPtr attr) -> bool
+		{
+		AttrTag new_tag = attr->Tag();
+
+		if ( new_tag == ATTR_DEPRECATED )
+			{
+			if ( ! attr->DeprecationMessage().empty() )
+				return false;
+
+			auto existing = Find(new_tag);
+			if ( existing && ! existing->DeprecationMessage().empty() )
+				return false;
+
+			return true;
+			}
+
+		return new_tag == ATTR_LOG || new_tag == ATTR_OPTIONAL || new_tag == ATTR_REDEF ||
+			new_tag == ATTR_BROKER_STORE_ALLOW_COMPLEX || new_tag == ATTR_RAW_OUTPUT ||
+			new_tag == ATTR_ERROR_HANDLER;
+		};
+
+	// Display a warning for duplicated tags on a type. &log tags are intentionally
+	// ignored here because duplicate log tags on record types are valid, and don't
+	// cause any significant breakage for other types.
+	if ( ! acceptable_duplicate_tag(attr) && Find(attr->Tag()) )
```

Comment:
* Can change order to short-circuit: if the attr isn't duplicated, no reason to additionally check if duplication is acceptable for that attr
* Also could pass along the result of the `Find()` to the lambda since the `ATTR_DEPRECATED` can end up doing the same `Find()` again otherwise

Reply:
> Can change order to short-circuit: if the attr isn't duplicated, no reason to additionally check if duplication is acceptable for that attr

It's probably hit-or-miss on whether the check for the tag type is faster/slower than the lookup for whether the tag exists. I don't feel strongly either way though.

> Also could pass along the result of the Find() to the lambda since the ATTR_DEPRECATED can end up doing the same Find() again otherwise

Good point.